### PR TITLE
Consolidate build process with GCC

### PR DESCRIPTION
**Pull request by gaearon**
This completely gets rid of Uglify in our build process. GCC provides better DCE for the FB bundles (e.g. the persistence path is now dead code eliminated in www, and mutation path is eliminated in CS). And this unifies our build process with how we already build UMD/CJS.

I forked GCC to add https://github.com/google/closure-compiler/pull/2707. Hopefully this gets merged but if not, it’s a pretty simple change to maintain. Then I recompiled the `google-closure-compiler-js` package and pointed Yarn lockfile to my fork (easier than forking another dependency layer up from there). I verified with Yarn folks it’s a supported (although a bit unobvious) way to point dependency to a fork (and this is pretty much what custom registry support relies on).

GCC doesn't preserve whitespace so I added Prettier after it for FB/RN bundles.

I also took an opportunity to clean up how we apply Rollup plugins in general, and set up a single chain instead of switches. This should make bundles more uniform across the build targets.

Some DEV sizes decreased because I started dropping per-file license headers. FB PROD sizes decreased because of better DCE and less whitespace.

RN DEV sizes increased because I did *not* add Closure to them whereas in the past they *did* use Uglify (for DEV). I'm not sure why it was desirable and it seemed better to make them un-Closure'd just like FB_DEV builds are. But I can change that if you want to. Mainly I just want to keep DEV bundles with comments, and Closure can't preserve them.

And that's about it! I hope to test and maybe land this on Thursday after shipping 16.1.

### CODE REVIEW

File: scripts/rollup/build.js [hunk 1]
```diff
@@ -416,88 +328,77 @@ function getPlugins(
   modulesToStub,
   featureFlags
 ) {
+  const findAndRecordErrorCodes = extractErrorCodes(errorCodeOpts);
   const shims = Modules.getShims(bundleType, entry, featureFlags);
-  const plugins = [
+  const isProduction = isProductionBundleType(bundleType);
+  const isInGlobalScope = bundleType === UMD_DEV || bundleType === UMD_PROD;
```
Comment (trueadm) [inline scripts/rollup/build.js]: I’m not sure this is the best name. Maybe `isUMDBundle`?

File: scripts/rollup/build.js [hunk 2]
```diff
@@ -506,9 +407,8 @@ function getPlugins(
           gzip,
         };
       },
-    })
-  );
-  return plugins;
+    }),
+  ].filter(Boolean);
```
Comment (trueadm) [inline scripts/rollup/build.js]: I’m not sure this is super obvious as to what’s happening?

File: scripts/rollup/build.js (hunk 1 above)
Comment (gaearon) [reply scripts/rollup/build.js]: There may be more bundle types where top-level things are declared in global scope. In theory. :-)

File: scripts/rollup/build.js (hunk 2 above)
Comment (gaearon) [reply scripts/rollup/build.js]: I dunno, seems like a common pattern with plugin systems?

screen shot 2017-11-08 at 08 11 31

Happy to change to something explicit if you prefer.

**Comment by gaearon** [comment]
>Almost all the node bundle sizes for production all seem to have regressed

I only see a ~20 byte difference. If that's what you're referring to, yes, it's expected because we no longer try to compress the license header, and it's consistent for DEV and PROD bundles.

**Comment by gaearon** [comment]
Added a second commit that doesn't change the build output, but breaks out the bundle header/footer logic into a separate file, and restructures it to be more explicit. This makes `build.js` really focused.

**Pull request merged**